src/main/java/com/opentable/kafka/util/LogProgressRestoreListener.java

```diff
@@ -0,0 +1,144 @@
+package com.opentable.kafka.util;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Arrays;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.stream.IntStream;
+
+import javax.annotation.concurrent.GuardedBy;
+
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.streams.KafkaStreams.State;
+import org.apache.kafka.streams.KafkaStreams.StateListener;
+import org.apache.kafka.streams.processor.StateRestoreListener;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class LogProgressRestoreListener implements StateRestoreListener, StateListener {
+    private static final Logger LOG = LoggerFactory.getLogger(LogProgressRestoreListener.class);
+
+    @GuardedBy("this")
+    private final Map<String, Partitions> restoreState = new LinkedHashMap<>();
+    private Instant lastPrint = Instant.now(), restoreStart = Instant.now();
+
+    private final StateListener delegate;
+
+    public LogProgressRestoreListener() {
+        this(null);
+    }
+
+    public LogProgressRestoreListener(StateListener delegate) {
+        this.delegate = delegate;
+    }
+
+    @Override
+    public void onChange(State newState, State oldState) {
+        if (delegate != null) {
+            delegate.onChange(newState, oldState);
+        }
+        if (newState == State.REBALANCING) {
+            LOG.info("Marking beginning of new restore operation");
+            restoreStart = Instant.now();
```

Comment:
This is the overall operation start. So maybe make it clearly different from onRestoreStart()
- beginningRestoreOperation ? or some such